bftengine/tests/bcstatetransfer/bcstatetransfer_tests.cpp

```diff
@@ -1246,6 +1248,47 @@ INSTANTIATE_TEST_CASE_P(BcStTest,
                                           BcStTestParamFixtureInput(100, 256),
                                           BcStTestParamFixtureInput(256, 100)), );
 
+// Validate that the source selector's primary awareness mechanism can be toggled on and off
+TEST_P(BcStTestParamFixture2, dstSourceSelectorPrimaryAwareness) {
+  auto [enable_primary_awareness, number_of_replacements] = GetParam();
+  targetConfig_.enableSourceSelectorPrimaryAwareness = enable_primary_awareness;
+  ASSERT_NFF(initialize());
+  std::once_flag once_flag;
+  ASSERT_NFF(dstStartRunningAndCollecting());
+  ASSERT_NFF(fakeSrcReplica_->replyAskForCheckpointSummariesMsg());
+  auto ss = stDelegator_->getSourceSelector();
+  const std::function<void(void)> f2 = [&]() {
+    std::call_once(once_flag, [&] {
+      std::unique_ptr<MessageBase> msg;
+      // Generate prePrepare messages to trigger source seletor to change the source to avoid primary.
+      ASSERT_NFF(msg = dataGen_->generatePrePrepareMsg(ss.currentReplica()));
+      for (uint16_t i = 1; i <= targetConfig_.minPrePrepareMsgsForPrimaryAwarness; i++) {
+        auto cmsg = make_shared<ConsensusMsg>(msg->type(), msg->senderId());
+        stateTransfer_->peekConsensusMessage(cmsg);
+      }
+    });
+  };
+  ASSERT_NFF(getMissingblocksStage(EMPTY_FUNC, f2));
+  const auto& sources = stDelegator_->getSourceSelector().getActualSources();
+  ASSERT_EQ(sources.size(), number_of_replacements);
```

Comment:
Not just total replacement should be verified but also the actual reason too which is due to source becoming primary. I believe there is such metric counter present not sure if its functional in dev branch.

Reply:
Thanks for pointing out.